src/program.cpp

```diff
@@ -105,829 +63,148 @@ program::program(const program& p) { assign(p); }
-    for(auto&& p : ins_vec)
-    {
-        double avg = common_average(p.second);
-        op_times[p.first->name()] += avg;
-        total_instruction_time += avg;
-    }
-    double calculate_overhead_time    = total_time - total_instruction_time;
-    double calculate_overhead_percent = calculate_overhead_time * 100.0 / total_time;
-
-    print_program(*this, [&](auto ins, const auto& names) {
-        print_instruction(std::cout, ins, names);
-
-        // skip return instruction
-        if(ins->name() == "@return")
-            return;
-
-        double avg     = common_average(ins_vec[ins]);
-        double percent = std::ceil(100.0 * avg / total_instruction_time);
-        os << ": " << avg << "ms, " << percent << "%";
-        os << std::endl;
-    });
-
-    os << std::endl;
-    os << "Summary:" << std::endl;
-    std::vector<std::pair<double, std::string>> op_times_sorted;
-    std::transform(op_times.begin(),
-                   op_times.end(),
-                   std::back_inserter(op_times_sorted),
-                   [](auto p) { return std::make_pair(p.second, p.first); });
-    std::sort(op_times_sorted.begin(), op_times_sorted.end(), std::greater<>{});
-    for(auto&& p : op_times_sorted)
-    {
-        auto&& name    = p.second;
-        double avg     = p.first;
-        double percent = std::ceil(100.0 * avg / total_instruction_time);
-        os << name << ": " << avg << "ms, " << percent << "%" << std::endl;
-    }
-
-    os << std::endl;
-
-    os << "Rate: " << rate << "/sec" << std::endl;
-    os << "Total time: " << total_time << "ms" << std::endl;
-    os << "Total instructions time: " << total_instruction_time << "ms" << std::endl;
-    os << "Overhead time: " << overhead_time << "ms"
-       << ", " << calculate_overhead_time << "ms" << std::endl;
-    os << "Overhead: " << std::round(overhead_percent) << "%"
-       << ", " << std::round(calculate_overhead_percent) << "%" << std::endl;
+    main_module.perf_report(os, n, std::move(params));
 }
 
-void program::debug_print() const { std::cout << *this << std::endl; }
-void program::debug_print(instruction_ref ins) const
-{
-    if(ins == this->end())
-    {
-        std::cout << "End instruction" << std::endl;
-        return;
-    }
-    if(not has_instruction(ins))
-    {
-        std::cout << "Instruction not part of program" << std::endl;
-        return;
-    }
-    std::stringstream ss;
-    print_program(*this, [&](auto x, const auto& names) {
-        if(x == ins)
-        {
-            print_instruction(std::cout, x, names);
-            std::cout << std::endl;
-        }
-    });
-}
+void program::debug_print() const { main_module.debug_print(); }
+void program::debug_print(instruction_ref ins) const { main_module.debug_print(ins); }
 void program::debug_print(const std::vector<instruction_ref>& inss) const
 {
-    for(auto ins : inss)
-        debug_print(ins);
-    std::cout << std::endl;
+    main_module.debug_print(inss);
```

Comment:
This should probably be removed from the program class, and only on the module class.